### DISCUSSION TITLE
Type aliases that cannot be assigned to other types than itself or it's basetype.

### DISCUSSION THREAD

**Discussion by thargol1** [dotnet/csharplang#8139] [Language Ideas]
## proposed syntax
``` csharp
using explicit Studentnumber = string;
using explicit Employeenumber = string;
```

## workings
Using explicit is a compiler only instruction that prohibits assignments to or from those types other than from the type itself or it's base type.

These assignments are ok
```csharp
Studentnumber a = "123456"; 
var b = (Studentnumber)"123456"
Studentnumber c = "123457";
a = c;
Employeenumber e = "12345";
a = (Studentnumber)e; // <-- not smart, but it will compile
string x = a;
x = e;
```
These assigments generate a compiler error:
```csharp
Studentnumber a = "123456";
Employeenumber e = a; // <-- error
Employeenumber f = (Studentnumber)"12345"; <-- error
```

These function calls generate a compiler error:
```csharp
Student GetStudent(Studentnumber s)
{
    ...
}

Employeenumber x = "12345";
var s = GetStudent(x) // <-- error
```

## Why
Many key-like types in applications are either string or integral type. Preventing those types from accidentily be assigned to each other could prevent errors during compile time.

I tried solutions like value structs. But they require a lot of extra code to make them work with JSON or EF.

## Lowering
This should be a compiler only function. 
```csharp
Studentnumber a = "123456"; 
var b = (Studentnumber)"123456"
Studentnumber c = "123457";
a = c;
Employeenumber e = "12345";
a = (Studentnumber)e;
```
Lowers to
```csharp
string a = "123456"; 
string b = "123456"
string c = "123457";
a = c;
string e = "12345";
a = e;

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/7642

**Comment by nyan-cat** [comment]
Opaque type aliases in C# is something we want since the last ice age

**Comment by FaustVX** [comment]
You will be able to do that with the `extension` proposal :
`explicit extension StudentNumber for string {}`
`explicit extension EmployeeNumber for string {}`

**Reply by 333fred** [reply]
No, I don't expect an explicit extension would prevent you from converting back to `string`.

**Reply by CyrusNajmabadi** [reply]
@333fred this is something i've proposed as part of that design.  I would *like* a way for the extension author to be able to opt-into exactly that sort of "explicit conversion" requirement.  

I consider this pretty important actually.  Something like an 'active extension' that only applies once you've verified it should be acceptable in that scenario (potentially throwing if it doesn't).

**Reply by FaustVX** [reply]
I'd still prefer to have a clearer distinction between the 2 `extension`, like `extension` and `role`.
To have syntax only in `role`, with some requirements: `role Email for string requires value.Contains('@');`
If possible, compile time check, otherwise, throws.

You can't implicitly convert from `string` to `Email` (each `string` is not always an `Email`), but `Email` to `string` yes (each `Email` is always a `string`).

**Reply by HaloFour** [reply]
IMO a syntax like that would make it challenging to support complex scenarios.  I'd rather a custom operator, such as an explicit conversion operator, or even a custom `is` operator.

```cs
public explicit extension Email for string {
    public static bool operator is(string value) {
        return value != null && value.Contains("@");
    }

    public static explicit operator Email(string value) {
        return value is Email email ? email : throw new Exception("Not an email!");
    }
}
```

**Reply by FaustVX** [reply]
Yes, but if you want complex scenarios, you can call your static method
```C"
role Email for string requires IsValid(value)
{
  static bool IsValid(string value)
  => value.Contains('@')
}
```
And if you want a simple constrain, it's very easy to write (and I hope that would be the main case)

**Reply by FaustVX** [reply]
Or, maybe, in the `requires` section, allows either a pattern (that can be compile check with a constant provided, otherwise runtime check) or a call to a method (which can't be compile check)

**Reply by HaloFour** [reply]
I just don't think it needs to be a special purpose syntax for extensions.  The language already has explicit conversion operators, and [custom/active patterns](https://github.com/dotnet/csharplang/issues/1047) are something that the team has mulled about as a general purpose feature.

**Reply by CyrusNajmabadi** [reply]
Agreed.  The cleanest way to do this, IMO, is to allow an explicit-extension to delcare an explicit-conversion-operator for this purpose (and maybe a custom `is T` operator).

**Reply by FaustVX** [reply]
Yes, I just wanted to have a simple and clear syntax for `role`

**Reply by CyrusNajmabadi** [reply]
`explicit extension Email for string` is that syntax.  It's simple, fits with 'implicit extensions' and is `explicit` by virtue that you have to explicitly use that name ("Email") to get the extension members vs just bringing them by using a namespace.

**Reply by HaloFour** [reply]
Either way, as long as the team is considering the possibility of having an "explicit extension" (or whatever it ends up being called) require a validation process I am happy.  I can see it being very useful for cases like `NumericString` or `CustomerId` where the latter requires a specific format, etc.

**Reply by CyrusNajmabadi** [reply]
Or any of the Json->Model wrappers that people might make ("Order", "Product", etc.).

**Reply by FaustVX** [reply]
@HaloFour Yes, it's just that _I_ prefer to have a better distinction between the 2.
At the end, I'll be happy whatever it'll be called, as long as at the final stage they have some validation process included.

**Reply by thargol1** [reply]
I like the syntax proposed by @HaloFour. It's clean and readable.

Would having an `is` operator also allow for stuff like this?:
```csharp
string a = GetInput();
if (a is Email email) Console.WriteLine($"valid {email}");
```

And would implicit conversion back to `string` still be possible? I don't like writing custom converters for EF, JSON, XML and other future packages that may come our way.

And would this be a compiler only option? So I can use it in old .Net framework 4.8 projects, where I now can use C#12 without problems.

**Reply by CyrusNajmabadi** [reply]
> I like the syntax proposed by @HaloFour. It's clean and readable.

That *is* the syntax for extensions :)  

> And would this be a compiler only option?

Unclear what this means.  We're discussing language changes here.  No language changes will work unless you get a compiler that supports that language version.

**Reply by thargol1** [reply]
With a 'compiler only option' I mean new syntax that doesn't require updates to the frameworks like .Net Core 9, but works with any supported version of .Net Core or .Net Framework.

**Reply by CyrusNajmabadi** [reply]
That's out of scope for hte lang design part of things.  Speaking to the team/compiler position: we can and will take advantage of new runtime features, and lang features have no guarantee of working downlevel *at all*.  If they do, great.  But you're in completely unsupported territory, and things may break at any moment, without any support.
